source/CIS/cis-5.rst

```diff
@@ -0,0 +1,599 @@
+
+``InternalNativeCurrencyTransferEvent``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A ``InternalNativeCurrencyTransferEvent`` SHALL be logged for every ``internalNativeCurrencyTransfer`` function invoke.
+
+The ``InternalNativeCurrencyTransferEvent`` is serialized as: First a byte with the value of 245, followed by the :ref:`CIS-5-CCDAmount` (``ccdAmount``), a :ref:`CIS-5-PublicKeyEd25519` (``from``), and the :ref:`CIS-5-PublicKeyEd25519` (``to``)::
+
+  InternalNativeCurrencyTransferEvent ::= (245: Byte) (ccdAmount: CCDAmount) (from: PublicKeyEd25519) (to: PublicKeyEd25519)
+
+``InternalCis2TokensTransferEvent``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A ``InternalCis2TokensTransferEvent`` SHALL be logged for every ``internalCis2TokensTransfer`` function invoke.
+
+The ``InternalCis2TokensTransferEvent`` is serialized as: First a byte with the value of 244, followed by the
+:ref:`CIS-5-TokenAmount` (``tokenAmount``), :ref:`CIS-5-TokenID` (``TokenID``),
+:ref:`CIS-5-ContractAddress` (``cis2TokenContractAddress``), a :ref:`CIS-5-PublicKeyEd25519` (``from``), and the :ref:`CIS-5-PublicKeyEd25519` (``to``)::
+
+  InternalCis2TokensTransferEvent ::= (244: Byte) (tokenAmount: TokenAmount) (tokenId: TokenID) (cis2TokenContractAddress: ContractAddress) (from: PublicKeyEd25519) (to: PublicKeyEd25519)
+
+
+
+.. _CIS-5-functions:
+
+Contract functions
+------------------
+
+A smart contract implementing this standard MUST export the following functions:
+
+- :ref:`CIS-5-functions-depositNativeCurrency`
+- :ref:`CIS-5-functions-depositCis2Tokens`
+- :ref:`CIS-5-functions-withdrawNativeCurrency`
+- :ref:`CIS-5-functions-withdrawCis2Tokens`
+- :ref:`CIS-5-functions-internalNativeCurrencyTransfer`
+- :ref:`CIS-5-functions-internalCis2TokensTransfer`
+- :ref:`CIS-5-functions-balanceOfNativeCurrency`
+- :ref:`CIS-5-functions-balanceOfCis2Tokens`
+
+
+.. _CIS-5-functions-depositNativeCurrency:
+
+``depositNativeCurrency``
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The function is payable and deposits/assigns the send CCDAmount (native currency) to a public key (``PublicKeyEd25519``).
+
+Parameter
+~~~~~~~~~
+
+The parameter is a ``PublicKeyEd25519``.
+
+See the serialization rules in :ref:`CIS-5-PublicKeyEd25519`.
+
+Requirements
+~~~~~~~~~~~~
+
+- The function MUST emit a ``DepositNativeCurrencyEvent``.
+
+.. _CIS-5-functions-depositCis2Tokens:
+
+``depositCis2Tokens``
+^^^^^^^^^^^^^^^^^^^^^
+
+This function SHOULD be called through the receive hook mechanism (:ref:`CIS-2-Receive-Hook-Function`)
+of a CIS-2 token contract. The function deposits/assigns the send CIS-2 token amount to a public key (``PublicKeyEd25519``).
+
+.. note::
+
+  If a use case wants to mint and deposit tokens to a public key in one transaction.
+  The CIS2 token has to have a mint function that calls this smart contract wallet ``depositCis2Tokens`` function via a hook mechanism.
```

Comment:
**Suggestion:** I don't think this note is relevant for the standard, how tokens are minted is not part of CIS-2 and, also it could be done in the opposite order, having a custom function in the CIS-5 contract, which calls a `mint` on the CIS-2 contract.